### PR TITLE
feat: Scroll shadow for new IDE toolbar

### LINKED ISSUE #37578: [Task]: Remove border bottom and add shadow for toolbar

### PR DESCRIPTION
**Pull request by albinAppsmith**
## Description

This PR have following changes to IDE toolbar
- Removed bottom border
- Added shadow on scroll of content

Fixes #37578

## Automation

/ok-to-test tags="@tag.Sanity"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11966309500>
> Commit: 8b6364a8a2b1717801ecd653bc37a48607a2c067
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11966309500&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Sanity`
> Spec:
> <hr>Fri, 22 Nov 2024 04:51:41 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Release Notes

- **New Features**
	- Introduced a new utility function `cn` for merging class names, including support for Tailwind CSS classes.
	- Enhanced the `Toolbar` component with scroll-based shadow effects for improved visual feedback.

- **Improvements**
	- Updated `Box` and `Flex` components to support ref forwarding, increasing their flexibility in usage.

- **Testing**
	- Added a comprehensive test suite for the `cn` function to ensure reliable class name merging behavior.

- **Dependencies**
	- Added `tailwind-merge` library to enhance styling capabilities.

### CODE REVIEW

File: app/client/packages/utils/src/cn/cn.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { type ClassValue, clsx } from "clsx";
+import { twMerge } from "tailwind-merge";
+
+function cn(...inputs: ClassValue[]) {
```
Comment (hetunandu) [inline app/client/packages/utils/src/cn/cn.ts:4]: What is `cn` ? Can we name it better? Why cant we use twMerge directly?

File: app/client/src/IDE/Structure/Toolbar.tsx [hunk 2]
```diff
@@ -1,19 +1,59 @@
...
+        }
+      };
+    }
+  }, []);
+
   return (
     <Flex
       alignItems="center"
-      borderBottom="1px solid var(--ads-v2-color-border-muted);"
+      className={cn("sticky top-0 transition-shadow duration-300 ease-in-out", {
```
Comment (hetunandu) [inline app/client/src/IDE/Structure/Toolbar.tsx:48]: Was this not possible with "classnames" or "clx" ? Why did we introduce a new library for this

File: app/client/src/IDE/Structure/Toolbar.tsx [hunk 3]
```diff
@@ -1,19 +1,59 @@
...
+      const adjacentElement = toolbarRef.current
+        .nextElementSibling as HTMLElement;
+
+      if (adjacentElement) {
+        adjacentElement.addEventListener("scroll", handleScroll, {
+          capture: true,
+        });
+      }
+
+      return () => {
```
Comment (hetunandu) [inline app/client/src/IDE/Structure/Toolbar.tsx:35]: This return (unsubscribe) is inside the `toolbarRef.current` check block and that is wrong

File: app/client/packages/utils/src/cn/cn.ts (hunk 1 above)
Comment (albinAppsmith) [reply app/client/packages/utils/src/cn/cn.ts:4]: https://theappsmith.slack.com/archives/C07C2UMN91T/p1727771273014209?thread_ts=1727759931.700239&cid=C07C2UMN91T

**Comment by albinAppsmith** [comment]
Closing this PR for now because of feasibility issues

**Pull request closed without merging**